[pytorch] Convert FORWARD_AUTOGRAD_REQ handler wait case to future.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31374 [pytorch] Convert FORWARD_AUTOGRAD_REQ handler wait case to future.**

This change converts the existing wait() functionality in the forward pass
to a future-based idiom.

The one complication is handling the contextId in the future handler, but
we introduce a new ContextPusher helper class temporarily push/restore the
contextId in that handler.

Differential Revision: [D19144795](https://our.internmc.facebook.com/intern/diff/D19144795/)

Commit:
[pytorch] Convert FORWARD_AUTOGRAD_REQ handler wait case to future.

This change converts the existing wait() functionality in the forward pass
to a future-based idiom.

The one complication is handling the contextId in the future handler, but
we introduce a new ContextPusher helper class temporarily push/restore the
contextId in that handler.

Differential Revision: [D19144795](https://our.internmc.facebook.com/intern/diff/D19144795/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit aca0d18:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=aca0d1851c944f10ef5290a0f9f8af8dcbca207c).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3981831) pytorch_linux_xenial_py3_6_gcc5_4_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76009012) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76009012))
<details>
<summary>
<code>Dec 17 19:45:21 RuntimeError: test_dist_autograd_spawn failed!</code>
</summary>

```
Dec 17 19:45:21 Ran 48 tests in 1327.516s 
Dec 17 19:45:21  
Dec 17 19:45:21 FAILED (failures=39, errors=1, skipped=2) 
Dec 17 19:45:21  
Dec 17 19:45:21 Generating XML reports... 
Dec 17 19:45:21 Traceback (most recent call last): 
Dec 17 19:45:21   File "test/run_test.py", line 456, in <module> 
Dec 17 19:45:21     main() 
Dec 17 19:45:21   File "test/run_test.py", line 449, in main 
Dec 17 19:45:21     raise RuntimeError(message) 
Dec 17 19:45:21 RuntimeError: test_dist_autograd_spawn failed! 
Dec 17 19:45:21 + cleanup 
Dec 17 19:45:21 + retcode=1 
Dec 17 19:45:21 + set +x 
Dec 17 19:45:21 =================== sccache compilation log =================== 
Dec 17 19:45:21 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Dec 17 19:45:21  
Dec 17 19:45:21 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 17 19:45:21 Compile requests                 0 
Dec 17 19:45:21 Compile requests executed        0 
Dec 17 19:45:21 Cache hits                       0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[pytorch] Convert FORWARD_AUTOGRAD_REQ handler wait case to future."


This change converts the existing wait() functionality in the forward pass
to a future-based idiom.

The one complication is handling the contextId in the future handler, but
we introduce a new ContextPusher helper class temporarily push/restore the
contextId in that handler.

Differential Revision: [D19144795](https://our.internmc.facebook.com/intern/diff/D19144795/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/autograd/context/container.h:
```diff
@@ -123,6 +123,15 @@ class TORCH_API DistAutogradContainer {
   int64_t max_id_;
 };
 
+class ContextPusher {
```
Add some docs for this class.

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -200,18 +200,38 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+              responseFuture->markCompleted(std::move(m));
+            }
+          });
+      return responseFuture;
```
What is the benefit of making this async vs just waiting? `processRpc` mostly just does CPU work

jjlilley replied:
Agree that it does largely do cpu work currently, and probably won't be as impactful as I hoped when I wrote up the diff.

fwiw, my original motivation here is simply to get rid of gratuitous synchronous-ness in this handler function; I've been trying to send out PRs (some bite-sized like this) and expedite reviews of PRs that help improve parallelism without throwing hundreds of mostly-waiting threads at the problem. Since long-term, zillions-of-threads is a recipe for memory thrash.

But... given the test failures (that I thought I had working locally when running yesterday, not sure what's going on), this might be lower priority, and might shelve this unless I can untangle a bit.

jjlilley commented:
Per prev comment, closing this PR temporarily until I get a chance to look at the failures more closely
